File: apps/backend/app/models/signature_event.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class SignatureEvent(Base):
    __tablename__ = "signature_events"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    patient_id: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    document_type: Mapped[str] = mapped_column(String(50), nullable=False, index=True)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="captured")
    signer_name: Mapped[str] = mapped_column(String(255), nullable=False)
    info_text_version: Mapped[str | None] = mapped_column(String(50), nullable=True)
    source: Mapped[str] = mapped_column(String(30), nullable=False, default="admin_test")
    note: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_by_user_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("users.id"), nullable=True
    )

    signed_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, nullable=False, index=True
    )

    # Nur für document_type='vp_antrag' relevant: hat die Krankenkasse den
    # Antrag bestätigt? Vom Büro im Admin-Web gesetzt. Wenn True zeigt der
    # Mobile-VP-Screen eine grüne "genehmigt"-Card statt "offen".
    approved_by_kk: Mapped[bool] = mapped_column(
        Boolean, default=False, nullable=False
    )
    approved_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    approved_note: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )

    asset = relationship(
        "SignatureAsset",
        back_populates="event",
        uselist=False,
        cascade="all, delete-orphan",
    )

File: apps/mobile/lib/core/models/signature_event.dart
enum DocumentType {
  leistungsnachweis,
  vpAntrag,
  pflegeumwandlung,
  betreuungsvertrag,
  pflegeantragHilfsmittel;

  String get apiValue {
    switch (this) {
      case DocumentType.leistungsnachweis:
        return 'leistungsnachweis';
      case DocumentType.vpAntrag:
        return 'vp_antrag';
      case DocumentType.pflegeumwandlung:
        return 'pflegeumwandlung';
      case DocumentType.betreuungsvertrag:
        return 'betreuungsvertrag';
      case DocumentType.pflegeantragHilfsmittel:
        return 'pflegeantrag_hilfsmittel';
    }
  }

  String get label {
    switch (this) {
      case DocumentType.leistungsnachweis:
        return 'Leistungsnachweis';
      case DocumentType.vpAntrag:
        return 'Verhinderungspflege';
      case DocumentType.pflegeumwandlung:
        return 'Pflegeumwandlung';
      case DocumentType.betreuungsvertrag:
        return 'Betreuungsvertrag';
      case DocumentType.pflegeantragHilfsmittel:
        return 'Pflegehilfsmittel';
    }
  }

  static DocumentType fromApiValue(String value) {
    switch (value) {
      case 'leistungsnachweis':
        return DocumentType.leistungsnachweis;
      case 'vp_antrag':
        return DocumentType.vpAntrag;
      case 'pflegeumwandlung':
        return DocumentType.pflegeumwandlung;
      case 'betreuungsvertrag':
        return DocumentType.betreuungsvertrag;
      case 'pflegeantrag_hilfsmittel':
        return DocumentType.pflegeantragHilfsmittel;
      default:
        throw ArgumentError('Unknown document type: $value');
    }
  }
}

class SignatureEvent {
  final int id;
  final int patientId;
  final DocumentType documentType;
  final String status;
  final String signerName;
  final String source;
  final String? note;
  final int? createdByUserId;
  final DateTime signedAt;
  final bool approvedByKk;
  final DateTime? approvedAt;
  final String? approvedNote;
  final DateTime createdAt;

  const SignatureEvent({
    required this.id,
    required this.patientId,
    required this.documentType,
    required this.status,
    required this.signerName,
    required this.source,
    this.note,
    this.createdByUserId,
    required this.signedAt,
    this.approvedByKk = false,
    this.approvedAt,
    this.approvedNote,
    required this.createdAt,
  });

  factory SignatureEvent.fromJson(Map<String, dynamic> json) {
    return SignatureEvent(
      id: json['id'] as int,
      patientId: json['patient_id'] as int,
      documentType: DocumentType.fromApiValue(json['document_type'] as String),
      status: json['status'] as String,
      signerName: json['signer_name'] as String,
      source: json['source'] as String,
      note: json['note'] as String?,
      createdByUserId: json['created_by_user_id'] as int?,
      signedAt: DateTime.parse(json['signed_at'] as String),
      approvedByKk: (json['approved_by_kk'] as bool?) ?? false,
      approvedAt: json['approved_at'] != null
          ? DateTime.parse(json['approved_at'] as String)
          : null,
      approvedNote: json['approved_note'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}
